File: examples/typescript/x402/clients/payForApi.ts
// Usage: pnpm tsx x402/clients/payForApi.ts

/**
 * Pay for an x402-protected API using a CDP-managed wallet.
 *
 * `CdpX402Client` auto-provisions a CDP Server Wallet (EOA) named
 * `"x402-client-wallet-1"` (override via `walletConfig.accountName`),
 * registers EVM and Solana payment schemes, and handles 402 responses
 * automatically. All credentials are read from environment variables.
 *
 * Setup:
 *   Set CDP_API_KEY_ID, CDP_API_KEY_SECRET, CDP_WALLET_SECRET in your .env
 *
 * Funding the wallet (Base Sepolia USDC):
 *   The example resolves and prints the client's wallet addresses upfront via
 *   `client.getAddresses()` — this is how you find which address to fund,
 *   since the wallet used for payment is otherwise managed internally. Fund
 *   it before paying, or set X402_FUND_FROM_FAUCET=true to auto-request USDC
 *   from the CDP faucet on startup. See the x402 examples README for funding
 *   options and env vars.
 */
import "dotenv/config";

import { CdpClient } from "@coinbase/cdp-sdk";
import { CdpX402Client } from "@coinbase/cdp-sdk/x402";
import { wrapFetchWithPayment } from "@x402/fetch";

const X402_PAID_API_URL =
  process.env.X402_API_URL ?? "https://x402.vercel.app/protected";

async function main() {
  // CdpX402Client lazily provisions its wallet on first payment and handles
  // 402 responses automatically. Calling `getAddresses()` eagerly provisions
  // it now, so we know which address to fund before paying.
  const client = new CdpX402Client();
  const { evmAddress, svmAddress } = await client.getAddresses();

  console.log("CDP-managed x402 client ready");
  console.log("  EVM address:", evmAddress);
  console.log("  Solana address:", svmAddress);
  console.log("  Fund the EVM address with USDC on Base Sepolia before making payments:");
  console.log("    CDP Faucet: https://portal.cdp.coinbase.com -> \"Onchain Tools\" -> \"Faucet\"\n");

  // Optional: top up the wallet straight from the CDP faucet. The same CDP
  // credentials power both the faucet and the x402 facilitator.
  if (process.env.X402_FUND_FROM_FAUCET === "true") {
    console.log("Requesting USDC from the CDP faucet...");
    try {
      const cdpClient = new CdpClient();
      const { transactionHash } = await cdpClient.evm.requestFaucet({
        address: evmAddress,
        network: "base-sepolia",
        token: "usdc",
      });
      console.log(`  Faucet tx: ${transactionHash}`);
      console.log("  Wait for it to confirm, then re-run without the flag to pay.\n");
      return;
    } catch {
      // The wallet may already be funded, or the project faucet limit was hit.
      console.warn("  Faucet request failed — you may already be funded, or hit the project limit.");
      console.warn("  Fund manually if needed: https://portal.cdp.coinbase.com -> \"Onchain Tools\" -> \"Faucet\"\n");
      // Fall through and attempt the payment anyway.
    }
  }

  const fetchWithPayment = wrapFetchWithPayment(globalThis.fetch, client);

  console.log(`Requesting: ${X402_PAID_API_URL}`);
  const response = await fetchWithPayment(X402_PAID_API_URL);

  if (!response.ok) {
    throw new Error(`Request failed: ${response.status} ${response.statusText}`);
  }

  console.log(`Payment succeeded — HTTP ${response.status}`);

  // The protected resource decides its own content type, so parse defensively
  // rather than assuming JSON (the x402.org demo endpoint returns HTML).
  const contentType = response.headers.get("content-type") ?? "";
  if (contentType.includes("application/json")) {
    console.log("Response:", JSON.stringify(await response.json(), null, 2));
  } else {
    const text = await response.text();
    console.log("Response:", text.length > 500 ? `${text.slice(0, 500)}… (${text.length} bytes)` : text);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/x402/clients/pay_for_api.py
# Usage: uv run python x402/clients/pay_for_api.py

"""Pay for an x402-protected API with a CDP-managed wallet (async, httpx).

The standard x402 Foundation client stack with one CDP swap: the signer is a CDP
API Key Wallet reached through eth_account's account interface via
``EvmLocalAccount``, so there is no private key to store. ``x402HttpxClient``
runs the 402 -> pay -> retry loop for you.

Python's CDP SDK has no ``CdpX402Client`` (unlike TypeScript), so the payment
schemes are registered explicitly.

Setup: set CDP_API_KEY_ID, CDP_API_KEY_SECRET and CDP_WALLET_SECRET in
examples/python/.env, then fund the printed address with Base Sepolia USDC, or
set CDP_FUND_FROM_FAUCET=true to self-fund on the first run.

Run: uv run python x402/clients/pay_for_api.py
"""

import asyncio
import os

from cdp import CdpClient
from cdp.evm_local_account import EvmLocalAccount
from dotenv import load_dotenv
from x402 import x402Client
from x402.http import x402HTTPClient
from x402.http.clients import x402HttpxClient
from x402.mechanisms.evm import EthAccountSigner
from x402.mechanisms.evm.exact import ExactEvmScheme
from x402.mechanisms.evm.upto import UptoEvmScheme

load_dotenv()

API_URL = os.getenv("X402_API_URL", "https://x402.vercel.app/protected")
NETWORK = "eip155:84532"  # Base Sepolia


async def main() -> None:
    async with CdpClient() as cdp:
        account = await cdp.evm.get_or_create_account(name="x402-client-wallet-1")
        # EthAccountSigner adapts the CDP account to the x402 signer protocol.
        signer = EthAccountSigner(EvmLocalAccount(account))
        print(f"Paying from {signer.address}")

        # The same CDP credentials power both the faucet and the facilitator.
        if os.getenv("CDP_FUND_FROM_FAUCET", "").lower() == "true":
            tx = await cdp.evm.request_faucet(
                address=signer.address, network="base-sepolia", token="usdc"
            )
            print(f"Requested USDC from the CDP faucet: {tx}")
            print("Wait for it to confirm, then re-run without the flag to pay.")
            return

        # TypeScript's CdpX402Client registers these schemes for you.
        payment_client = x402Client()
        payment_client.register(NETWORK, ExactEvmScheme(signer))
        payment_client.register(NETWORK, UptoEvmScheme(signer))

        async with x402HttpxClient(payment_client) as http:
            response = await http.get(API_URL)
            await response.aread()

        print(f"HTTP {response.status_code} ({len(response.content)} bytes)")

        # PAYMENT-RESPONSE carries the settlement receipt for the retried request.
        # A 200 can arrive without one if the endpoint did not charge, so treat a
        # missing receipt as information rather than an error.
        http_client = x402HTTPClient(payment_client)
        try:
            settled = http_client.get_payment_settle_response(response.headers.get)
            print(f"Settled tx: {settled.transaction}")
        except ValueError:
            print("No settlement receipt on the response — nothing was charged.")


if __name__ == "__main__":
    asyncio.run(main())
